kube-runtime/src/watcher.rs

```diff
@@ -309,6 +429,70 @@ pub fn watcher<K: Resource + Clone + DeserializeOwned + Debug + Send + 'static>(
+///
+/// Errors from the underlying watch are propagated, after which the stream will go into recovery mode on the next poll.
+/// You can apply your own backoff by not polling the stream for a duration after errors.
+/// Keep in mind that some [`TryStream`](futures::TryStream) combinators (such as
+/// [`try_for_each`](futures::TryStreamExt::try_for_each) and [`try_concat`](futures::TryStreamExt::try_concat))
+/// will terminate eagerly as soon as they receive an [`Err`].
+///
+/// This is intended to provide a safe and atomic input interface for a state store like a [`reflector`].
+/// Direct users may want to flatten composite events via [`WatchStreamExt`]:
+///
+/// ```no_run
+/// use kube::{
+///   api::{Api, ListParams, ResourceExt}, Client,
+///   runtime::{watcher, watcher::watch_metadata, WatchStreamExt}
+/// };
+/// use k8s_openapi::api::core::v1::Pod;
+/// use futures::{StreamExt, TryStreamExt};
+/// #[tokio::main]
+/// async fn main() -> Result<(), watcher::Error> {
+///     let client = Client::try_default().await.unwrap();
+///     let pods: Api<Pod> = Api::namespaced(client, "apps");
+///
+///     watch_metadata(pods, ListParams::default()).applied_objects()
+///         .try_for_each(|p| async move {
+///          println!("Applied: {}", p.name_any());
+///             Ok(())
+///         })
+///         .await?;
+///    Ok(())
+/// }
+/// ```
+/// [`WatchStreamExt`]: super::WatchStreamExt
+/// [`reflector`]: super::reflector::reflector
+/// [`Api::watch`]: kube_client::Api::watch
+///
+/// # Recovery
+///
+/// The stream will attempt to be recovered on the next poll after an [`Err`] is returned.
+/// This will normally happen immediately, but you can use [`StreamBackoff`](crate::utils::StreamBackoff)
+/// to introduce an artificial delay. [`default_backoff`] returns a suitable default set of parameters.
+///
+/// If the watch connection is interrupted, then `watcher` will attempt to restart the watch using the last
+/// [resource version](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes)
+/// that we have seen on the stream. If this is successful then the stream is simply resumed from where it left off.
+/// If this fails because the resource version is no longer valid then we start over with a new stream, starting with
+/// an [`Event::Restarted`]. The internals mechanics of recovery should be considered an implementation detail.
+
+pub fn watch_metadata<K: Resource + Clone + DeserializeOwned + Debug + Send + 'static>(
+    api: Api<K>,
+    list_params: ListParams,
+) -> impl Stream<Item = Result<Event<PartialObjectMeta>>> + Send {
+    futures::stream::unfold(
+        (api, list_params, State::Empty),
+        |(api, list_params, state)| async {
+            let (event, state) = step_metadata(&api, &list_params, state).await;
+            Some((event, (api, list_params, state)))
+        },
+    )
+}
+
```

Comment:
A thing I still wish for is some kind of test (but that is a tall ask atm as we [don't even test watcher much](https://app.codecov.io/gh/kube-rs/kube/blob/main/kube-runtime/src/watcher.rs) - some indirect integration/e2e uses of `await_condition` only). At the very least we could adapt `dynamic_watcher` to take an evar override to use `watch_metadata` instead of `watcher` and rely on compile time verification of that.

It's also possibly useful to have some sanity that the output type of `watch_metadata` and `watcher` stays compatible when given an arbitrary `K: Resource` could also be wise (given we ultimately want to inject this into further controller machinery down the line). They are compatible atm (given `PartialObjectMeta` has the same impls as is required by `WatchStreamExt`):
```rust
impl Stream<Item = Result<Event<PartialObjectMeta>>> + Send
impl Stream<Item = Result<Event<K>>> + Send
```
but maybe it makes sense to have a type alias or compile time checks for impls to prevent drift 🤔

Reply:
Suggestions make a lot of sense! I looked for some integration tests here but noticed we don't test the watcher. 

Updated `dynamic_watcher`, might look a bit... wonky with the evar override but I didn't really want to refactor more there, seems like the code should be easy to understand even if we duplicate a few things. If you see the code playing out differently there, let me know, happy to make as many changes as ya want.

For compiler assertions, I don't have a lot of experience (although I've used something similar to type check some linkerd proxy stacks once or twice). A type alias would be amazing but it's still unstable (`impl` in type aliases, that is) -- I wanted to introduce one early on, got fatigue from writing so many `impl Stream`s.

What I understood is that we want to make sure the return types align regardless of what `K: Resource` is. Based on your suggestion, I added a test to check this. Since we always run the tests we should spot any compilation errors early on. I found that the runtime `mod` file already had a typecheck so I piggybacked on that.

Let me know if the check is what you had in mind and if my comments on the functions make sense :)

Reply:
> Updated dynamic_watcher, might look a bit... wonky with the evar override but I didn't really want to refactor more there, seems like the code should be easy to understand even if we duplicate a few things. If you see the code playing out differently there, let me know, happy to make as many changes as ya want.

ah. i read this after pointing out some of the refactor potentials 🙃

Reply:
And yeah, you've got a good compile test there. I haven't gotten a lot of experience with this type of test either, it just felt like a useful invariant to protect, and this looks like it should do it 👍